components/mbedtls/esp_crt_bundle/esp_crt_bundle.c

```diff
@@ -1,89 +1,244 @@
+        ret = MBEDTLS_ERR_PK_TYPE_MISMATCH;
         goto cleanup;
     }
 
     md_info = mbedtls_md_info_from_type(child->MBEDTLS_PRIVATE(sig_md));
-    if ( (ret = mbedtls_md( md_info, child->tbs.p, child->tbs.len, hash )) != 0 ) {
-        ESP_LOGE(TAG, "Internal mbedTLS error %X", ret);
+
+    if( unlikely( md_info == NULL) ) {
+        ESP_LOGE(TAG, "Unknown message digest");
+        ret = MBEDTLS_ERR_X509_FEATURE_UNAVAILABLE;
         goto cleanup;
     }
 
-    if ( (ret = mbedtls_pk_verify_ext( child->MBEDTLS_PRIVATE(sig_pk), child->MBEDTLS_PRIVATE(sig_opts), &parent.pk,
-                                       child->MBEDTLS_PRIVATE(sig_md), hash, mbedtls_md_get_size( md_info ),
-                                       child->MBEDTLS_PRIVATE(sig).p, child->MBEDTLS_PRIVATE(sig).len )) != 0 ) {
+    {
 
-        ESP_LOGE(TAG, "PK verify failed with error %X", ret);
-        goto cleanup;
+        const unsigned char md_size = mbedtls_md_get_size(md_info);
+#if ESP_CRT_USE_VLA
+        unsigned char hash[md_size];
+#else
+        unsigned char hash[MBEDTLS_MD_MAX_SIZE];
+#endif
+
+        if ( unlikely( (ret = mbedtls_md( md_info, child->tbs.p, child->tbs.len, hash)) != 0) ) {
+            ESP_LOGE(TAG, "MD failed with error %X", ret);
+            goto cleanup;
+        }
+
+        if ( unlikely( (ret = mbedtls_pk_verify_ext( child->MBEDTLS_PRIVATE(sig_pk), child->MBEDTLS_PRIVATE(sig_opts), &pubkey,
+                                                     child->MBEDTLS_PRIVATE(sig_md), hash, md_size,
+                                                     child->MBEDTLS_PRIVATE(sig).p, child->MBEDTLS_PRIVATE(sig).len )) != 0) ) {
+
+            ESP_LOGE(TAG, "PK verify failed with error %X", ret);
+            goto cleanup;
+        }
     }
 cleanup:
-    mbedtls_x509_crt_free(&parent);
+    mbedtls_pk_free(&pubkey);
 
     return ret;
 }
 
+static size_t min(const size_t s1, const size_t s2) {
+    return s1 < s2 ? s1 : s2;
+}
+
+static cert_t esp_crt_find_cert(const unsigned char* const issuer, const size_t issuer_len)
+{
+    if (unlikely( issuer == NULL || issuer_len == 0 )) {
+        return NULL;
+    }
+
+    int start = 0;
+    int end = esp_crt_get_certcount(s_crt_bundle.bundle)-1;
+
+    /* Look for the certificate using binary search on subject name */
+    while (start <= end) {
+
+        const int middle = (start + end) / 2;
+
+        cert_t cert = esp_crt_get_cert(s_crt_bundle.bundle, middle);
+
+        const size_t cert_name_len = esp_crt_get_name_len(cert);
+
+        // Issuers are in DER encoding, with lengths encoded in the content; if valid DER, differing lengths
+        // are reflected in differing content.
+        // Still, we won't try to memcmp beyond the given length:
+        int cmp_res = memcmp(issuer, esp_crt_get_name(cert), min(issuer_len, cert_name_len) );
+
+        if ( unlikely( cmp_res == 0 ) ) {
+            cmp_res = (int)issuer_len - cert_name_len;
+            if( likely( cmp_res >= 0 ) ) {
```

Comment:
Hi @BitsForPeople, could you please explain why we allow the issuer name to be longer than the found certificate name? Do you have any example wherein this would be needed?

We think due to this, an incorrect certificate could get matched if the name of another certificate is a prefix of the name of the certificate we're looking for. What do you think?

Reply:
Hi @Harshal5.
I'm actually not sure anymore why I did a >= 0 here, t.b.h. Specifically, I can't remember seeing a case where this would be needed.
Reviewing it now, I believe you are right: If one cert's name would be a prefix of another cert's name in a bundle, we might (in theory) pick the wrong cert.
I think changing it to == 0 should work.

Sorry for seemingly not getting back to this PR until now. I spent a few hours trying to fix the issue of the unrelated mbedtls files in the PR but could not for the life of me get my working tree back to a consistent state with a clean PR.
Is there something you can do with the PR on your end?